src/migrate.c

```diff
@@ -0,0 +1,156 @@
+#include <string.h>
+
+#include "redisraft.h"
+
+int validSlotMagic(RedisRaftCtx *rr, unsigned int slot, int magic)
+{
+    /* validSlot will have already validated that slot is importing */
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    for (unsigned int i = 0; i < sg->slot_ranges_num; i++) {
+        ShardGroupSlotRange *sr = &sg->slot_ranges[i];
+        if (sr->start_slot <= slot && sr->end_slot >= slot && sr->magic == magic) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+int validSlotTerm(RedisRaftCtx *rr, int slot, raft_term_t term)
+{
+    ShardingInfo *si = rr->sharding_info;
+
+    if (si->max_importing_term[slot] <= term) {
+        si->max_importing_term[slot] = term;
+        return 1;
+    }
+
+    return 0;
+}
+
+int validSlot(RedisRaftCtx *rr, int slot)
+{
+    ShardGroup *sg = rr->sharding_info->importing_slots_map[slot];
+
+    if (sg && sg->local) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void importKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_IMPORT_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    ImportKeys import_keys = {0};
+    RedisModule_Assert(RaftRedisDeserializeImport(&import_keys, entry->data, entry->data_len) == RR_OK);
+    RedisModule_Assert(import_keys.num_keys > 0);
+
+    // FIXME: validate no cross slot migration at append time
+    int slot = KeyHashSlotRedisString(import_keys.key_names[0]);
+
+    if (!validSlot(rr, slot)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR not an importing slot");
+        }
+        goto exit;
+    }
+
+    if (!validSlotTerm(rr, slot, import_keys.term)) {
+        // this is acceptable (suspended old leader
+        LOG_WARNING("ignoring import keys as old term");
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid term");
+        }
+        goto exit;
+    }
+
+    if (!validSlotMagic(rr, slot, import_keys.magic)) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR invalid magic");
+        }
+        goto exit;
+    }
+
+    RedisModuleString * zero = RedisModule_CreateString(rr->ctx, "0", 1);
+
+    for (size_t i = 0; i < import_keys.num_keys; i++) {
+        RedisModuleString * temp[3];
+        temp[0] = import_keys.key_names[i];
+        temp[1] = zero;
+        temp[2] = import_keys.key_serialized[i];
+
+        enterRedisModuleCall();
+        RedisModuleCallReply *reply = RedisModule_Call(rr->ctx, "restore", "v", temp, 3);
```

Comment:
I think we should check error cases here, just in case.

Reply:
added an assert that it didn't fail and wraped in an existence check to not restore when key already exists.